### PR TITLE
Make head Postings only return series in time range

### PR DESCRIPTION
**Pull request by roidelapluie**
Reworks and closes #6458

### CODE REVIEW

File: tsdb/head.go [hunk 1]
```diff
@@ -1336,7 +1336,30 @@ func (h *headIndexReader) LabelNames() ([]string, error) {
 func (h *headIndexReader) Postings(name string, values ...string) (index.Postings, error) {
 	res := make([]index.Postings, 0, len(values))
 	for _, value := range values {
-		res = append(res, h.head.postings.Get(name, value))
+		p := h.head.postings.Get(name, value)
+		if h.maxt > h.head.maxTime && h.mint < h.head.minTime {
+			res = append(res, p)
+			continue
+		}
```
Comment (roidelapluie) [inline tsdb/head.go]: @brian-brazil I changed this compared to your first pr.

**Comment by roidelapluie** [comment]
benchmark                                  old ns/op      new ns/op      delta
BenchmarkHeadSeries/1of1000000-8           1182394114     111903957      -90.54%
BenchmarkHeadSeries/10of1000000-8          1161618692     111403226      -90.41%
BenchmarkHeadSeries/100of1000000-8         1169758844     112122961      -90.41%
BenchmarkHeadSeries/1000of1000000-8        1175302734     113658484      -90.33%
BenchmarkHeadSeries/10000of1000000-8       1191381454     123592113      -89.63%
BenchmarkHeadSeries/100000of1000000-8      1186973443     237907048      -79.96%
BenchmarkHeadSeries/1000000of1000000-8     1358181144     1374416302     +1.20%

benchmark                                  old allocs     new allocs     delta
BenchmarkHeadSeries/1of1000000-8           4000055        28             -100.00%
BenchmarkHeadSeries/10of1000000-8          4000073        87             -100.00%
BenchmarkHeadSeries/100of1000000-8         4000253        630            -99.98%
BenchmarkHeadSeries/1000of1000000-8        4002053        6036           -99.85%
BenchmarkHeadSeries/10000of1000000-8       4020053        60054          -98.51%
BenchmarkHeadSeries/100000of1000000-8      4200053        600074         -85.71%
BenchmarkHeadSeries/1000000of1000000-8     6000053        6000053        +0.00%

benchmark                                  old bytes     new bytes     delta
BenchmarkHeadSeries/1of1000000-8           229192185     2488          -100.00%
BenchmarkHeadSeries/10of1000000-8          229193337     5568          -100.00%
BenchmarkHeadSeries/100of1000000-8         229204857     35536         -99.98%
BenchmarkHeadSeries/1000of1000000-8        229320271     345104        -99.85%
BenchmarkHeadSeries/10000of1000000-8       230472057     3894672       -98.31%
BenchmarkHeadSeries/100000of1000000-8      241992057     40511632      -83.26%
BenchmarkHeadSeries/1000000of1000000-8     357192058     357192058     +0.00%

**Comment by roidelapluie** [comment]
```
BenchmarkHeadSeries/1of1000000-8           1182394114     111477804      -90.57%
BenchmarkHeadSeries/10of1000000-8          1161618692     111178736      -90.43%
BenchmarkHeadSeries/100of1000000-8         1169758844     112278785      -90.40%
BenchmarkHeadSeries/1000of1000000-8        1175302734     112791163      -90.40%
BenchmarkHeadSeries/10000of1000000-8       1191381454     122281323      -89.74%
BenchmarkHeadSeries/100000of1000000-8      1186973443     240413878      -79.75%
BenchmarkHeadSeries/1000000of1000000-8     1358181144     1387995740     +2.20%

benchmark                                  old allocs     new allocs     delta
BenchmarkHeadSeries/1of1000000-8           4000055        28             -100.00%
BenchmarkHeadSeries/10of1000000-8          4000073        87             -100.00%
BenchmarkHeadSeries/100of1000000-8         4000253        630            -99.98%
BenchmarkHeadSeries/1000of1000000-8        4002053        6036           -99.85%
BenchmarkHeadSeries/10000of1000000-8       4020053        60054          -98.51%
BenchmarkHeadSeries/100000of1000000-8      4200053        600074         -85.71%
BenchmarkHeadSeries/1000000of1000000-8     6000053        6000053        +0.00%

benchmark                                  old bytes     new bytes     delta
BenchmarkHeadSeries/1of1000000-8           229192185     2488          -100.00%
BenchmarkHeadSeries/10of1000000-8          229193337     5568          -100.00%
BenchmarkHeadSeries/100of1000000-8         229204857     35536         -99.98%
BenchmarkHeadSeries/1000of1000000-8        229320271     345104        -99.85%
BenchmarkHeadSeries/10000of1000000-8       230472057     3894672       -98.31%
BenchmarkHeadSeries/100000of1000000-8      241992057     40511632      -83.26%
BenchmarkHeadSeries/1000000of1000000-8     357192058     357192058     +0.00%
```

File: tsdb/block.go [hunk 2]
```diff
@@ -112,7 +112,7 @@ type ChunkReader interface {
 // BlockReader provides reading access to a data block.
 type BlockReader interface {
 	// Index returns an IndexReader over the block's data.
```
Comment (brian-brazil) [inline tsdb/block.go]: Probably needs an update. Note that there can be false positives.

File: tsdb/head_bench_test.go [hunk 3]
```diff
@@ -48,3 +49,35 @@ func BenchmarkHeadStripeSeriesCreateParallel(b *testing.B) {
 		}
 	})
 }
+
+func BenchmarkHeadSeries(b *testing.B) {
```
Comment (brian-brazil) [inline tsdb/head_bench_test.go]: This is already committed as BenchmarkQuerierSelect

File: tsdb/head.go [hunk 4]
```diff
@@ -750,7 +750,7 @@ type rangeHead struct {
 	mint, maxt int64
 }
 
-func (h *rangeHead) Index() (IndexReader, error) {
+func (h *rangeHead) Index(mint, maxt int64) (IndexReader, error) {
 	return h.head.indexRange(h.mint, h.maxt), nil
```
Comment (brian-brazil) [inline tsdb/head.go]: I'd expect the arguments to be used to limit these.

File: tsdb/head.go [hunk 5]
```diff
@@ -1334,9 +1334,33 @@ func (h *headIndexReader) LabelNames() ([]string, error) {
 
 // Postings returns the postings list iterator for the label pairs.
 func (h *headIndexReader) Postings(name string, values ...string) (index.Postings, error) {
+	mint, maxt := h.head.MinTime(), h.head.MaxTime()
 	res := make([]index.Postings, 0, len(values))
 	for _, value := range values {
-		res = append(res, h.head.postings.Get(name, value))
+		p := h.head.postings.Get(name, value)
+		if h.maxt >= maxt && h.mint <= mint {
```
Comment (brian-brazil) [inline tsdb/head.go]: A comment here would help, took me a bit to figure out what this was doing.

File: tsdb/block.go (hunk 2 above)
Comment (roidelapluie) [reply tsdb/block.go]: Which false positive?

Comment (brian-brazil) [reply tsdb/block.go]: Firstly you're not implementing this for blocks, secondly that a range overlaps a min/max doesn't mean it has samples there.

**Comment by brian-brazil** [comment]
I think we still need some additional tests here around Select.

**Comment by roidelapluie** [comment]
yeah especially since I am still a beginner in TSDB :)

**Comment by brian-brazil** [comment]
There's also subtly different semantics around open/closed time ranges between PromQL and blocks.

**Comment by roidelapluie** [comment]
@brian-brazil is that the kind of test you expect?

File: tsdb/head_test.go [hunk 6]
```diff
@@ -1315,3 +1315,74 @@ func TestAddDuplicateLabelName(t *testing.T) {
...
+	testutil.Ok(t, err)
+	defer h.Close()
+	app := h.Appender()
+	numSeries := 800
+	for i := 100; i < numSeries; i++ {
+		app.Add(labels.FromStrings(fmt.Sprintf("foo%d", i), "bar"), int64(i), 0)
+	}
+	testutil.Ok(t, app.Commit())
+
+	cases := []struct {
```
Comment (brian-brazil) [inline tsdb/head_test.go]: You should also test 99, 101, 798, and 801.

File: tsdb/head_test.go [hunk 7]
```diff
@@ -1315,3 +1315,74 @@ func TestAddDuplicateLabelName(t *testing.T) {
...
+
+	for i, c := range cases {
+		matcher := labels.MustNewMatcher(labels.MatchEqual, "", "")
+		q, err := NewBlockQuerier(h, c.mint, c.maxt)
+		testutil.Ok(t, err)
+
+		count := 0
+		ss, err := q.Select(matcher)
+		testutil.Ok(t, err)
+		for ss.Next() {
```
Comment (brian-brazil) [inline tsdb/head_test.go]: Don't just test the number of series, check the number of samples matches too.

Comment (roidelapluie) [reply tsdb/head_test.go]: done. Added better tests.

**Comment by roidelapluie** [comment]
Rebased, squashed + added benchmarks to my git commit.

File: tsdb/head.go [hunk 8]
```diff
@@ -763,8 +763,15 @@ func NewRangeHead(head *Head, mint, maxt int64) *RangeHead {
 	}
 }
 
-func (h *RangeHead) Index() (IndexReader, error) {
-	return h.head.indexRange(h.mint, h.maxt), nil
+func (h *RangeHead) Index(mint, maxt int64) (IndexReader, error) {
+	// rangeHead guarantees that the series returned are within its range.
+	if mint < h.mint {
```
Comment (brian-brazil) [inline tsdb/head.go]: This logic seems the wrong way around.

Comment (roidelapluie) [reply tsdb/head.go]: That is cropping the Index mint and maxt to the RangeHead's mint and maxt. It seems right to me. This seems to be the purpose of the RangeHead.

Comment (brian-brazil) [reply tsdb/head.go]: Ah yes

**Comment by brian-brazil** [comment]
:+1: 

I presume the benchmarks still look about the same after all the changes?

**Comment by roidelapluie** [comment]
Benchmarks are in both commit messages:

https://github.com/prometheus/prometheus/pull/6676/commits

You had -90%, I have -70%. There have been changes in TSDB since your original commit, not sure if related.

**Comment by roidelapluie** [comment]
mmh no benchmarks are both 90%

I guess I need to compare yours BenchmarkHeadSeries with BenchmarkQuerierSelect/SortedHead.

We have -70% in BenchmarkQuerierSelect/Head.

**Comment by brian-brazil** [comment]
No regression then, and we're now actually faster when all series are returned (which is surprising).

**Pull request merged**